### ISSUE TITLE
cylc run <abs-src-dir> - bug or feature?

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc#2749]
Standard/intended way to run a suite:
```
$ cylc register <NAME> <PATH>
$ cylc run <NAME>
# => suite run directory is $HOME/cylc-run/<NAME>/
$ cylc ping <NAME>  # etc.
```
@kinow's experiments with Docker have revealed that we can do this:
```
$ cylc run <PATH>
# => suite run directory is <PATH>
$ cylc ping <PATH>  # etc.
```
where PATH is the absolute path of the suite source directory.  Everything seems to work OK, except (e.g.) that `cylc scan` has to be forced to port scan because the filesystem scan under `$HOME/cylc-run` won't find the suite. The abs path works as the suite name for targeting the server.

@matthewrmshin - this strikes me as somewhat surprising ... is it known/intended behaviour that we should document, or an accident that we should prevent from working?

**Comment by hjoliver** [comment]
This mode of running would presumably screw up other things, such as Rose Bush / `cylc review` ...

**Comment by kinow** [comment]
Oh, interesting! I'll be lurking on this thread to see if I can learn one thing or two more about Cylc.

I'm using this same approach when debugging in PyDev, and - if memory serves me well - I learned it from the CUG... possibly from this part ? https://cylc.github.io/cylc/html/single/cug-html.html#7.11.1

```
$ cylc run tut/oneoff/basic 
```

Though I am not sure if relative/absolute paths are treated in the same way.

**Comment by hjoliver** [comment]
@kinow - suite names are used rather than path-to-suite-source-directory (because e.g. the path may be verbose, and is not really relevant when targeting a running suite server program):
```
$ cylc reg my-suite /path/to/suite/source/dir
# Creates run directory $HOME/cylc-run/my-suite
#   (in which `.service/source` is a symlink to the source dir)
$ cylc run my-suite
$ cylc stop my-suite
```
but names can be hierarchical like directory trees, and that gets reflected in the run directory tree:
```
$ cylc reg foo/bar/baz /path/to/suite/run/dir
# creates $HOME/cylc-run/foo/bar/baz/...
$ cylc run foo/bar/baz
...
```

Your example above works only if `tut/oneoff/basic` has been registered as a name.

If it is a relative path to the suite source directory, it fails as expected because the standard run directory has not been created via registration.

The surprising thing (new to me, or else I've forgotten!) is that it works if an absolute path is given to an unregistered suite (and the run dir ends up in the source dir as described above) ... which is the case for your recommended `docker run` command.

**Comment by kinow** [comment]
Oooohh, that makes sense now! And I just saw that the CUG has a command to register the suite before running it with the relative path. Thanks @hjoliver !!!

**Comment by hjoliver** [comment]
I suppose it is easy to confuse a hierarchical suite name with a relative path, especially as the name actually ends up as the the run directory path under `~/cylc-run/`.   Most of us are running suites via `rose suite-run` (don't worry about this for the moment) which does not use hierarchical suite names, so the confusion doesn't really arise in that case.

**Comment by hjoliver** [comment]
(this may be back-compatibility for the original in-place cylc run mode, where the source directory was not linked to the run directory ... @matthewrmshin will probably know - he refactored this part of the system some time ago...)

**Comment by matthewrmshin** [comment]
Maybe a bad side effect. I'll take another look tomorrow.

**Comment by matthewrmshin** [comment]
(Mostly harmless for normal users, but can get a bit strange.)

**Comment by hjoliver** [comment]
Yeah, we should probably disable this.  I can't think of a good reason to allow it.

**Comment by TomekTrzeciak** [comment]
This looks very interesting and I wish I knew about it earlier. I can think of one very good use case for this - running subsuites (#171, #2241). I was always missing a makefile-like mode in Cylc, where I could just run the suite in place (this issue) and without daemonizing it (`cylc run --no-detach`). It seems this was possible all along :smiley:. Could we get this promoted from a bug to a feature?

**Comment by hjoliver** [comment]
@TomekTrzeciak - I take it you would like the sub-suite run directory to end up in the source location, not under `~/cylc-run`, so it is not sufficient to do `cylc register` then `cylc run --no-detach` for the sub-suite?

We might need to think about implications ... e.g. the sub-suite output would be more or less hidden (from Rose Bush, for example).

**Comment by TomekTrzeciak** [comment]
@hjoliver - yes, I'm after an easy way to point sub-suite run directory somewhere else than `~/cylc-run`, e.g. ramdisk, and do something with the logs afterwords (like `cat` them all together in suite shutdown handler). I think after #2759 this should be still possible by explicitly setting `CYLC_SUITE_RUN_DIR` for the sub-suite (I haven't tried it though).

I guess Rose Bush would not be an issue for this, but as you say, there might be other implications to consider. All I'm asking here is that this mode of operation is given some further consideration rather than being treated outright as a bug. It might prove useful yet. (When I was first introduced to Cylc, my first expectation was that `cylc run /path/to/config/file.rc` would be the default way of executing the suite, with registration, daemonizing, etc. being taken care of by additional options or wrappers like Rose.)

**Comment by hjoliver** [comment]
Well, I'm actually a long-time fan of sub-suites, and got somewhat talked out of it because the housekeeping and "too many suites to monitor" problems.  So, anyway, I am happy to pursue this as you seem to have a good use case for it.

The reason for name registration, as I recall, was simply that the config file path doesn't make much sense as a way of targeting the network interface of a running suite server program, once you have started it up.

**Comment by TomekTrzeciak** [comment]
I think sub-suites, or perhaps even more ambitiously something like peer-to-peer distributed suites, could be a generic and potent idea forming the base to solve many problems. Things like job arrays, multidimensional and/or custom cycling modes, dynamic dependencies, etc. could all be potentially built on top of them. While I don't have any concrete ideas yet on how this could look like, I think it is worth exploring and experimenting in this space.

**Comment by hjoliver** [comment]
Sounds brilliant!  The sub-suite thing of course becomes more practical once we have the meta-scheduler abstracted out as kernel that can be applied in other contexts (as per @oliver-sanders' ideas) ... such as inside a task, but without being attached to the current verbose run directory structure etc.  

Can you elaborate on what you mean by "peer-to-peer distributed suites"? - just to add to the mix of ideas we want to address once the GUI+Python-3 crisis is dealt with.

**Comment by hjoliver** [comment]
> yes, I'm after an easy way to point sub-suite run directory somewhere else than ~/cylc-run, e.g. ramdisk, and do something with the logs afterwords (like cat them all together in suite shutdown handler). 

In the short term I think this sort of thing should be quite feasible, if @cylc/core don't see any major reasons why not.  In which case we should re-open #171 ...

**Comment by oliver-sanders** [comment]
I too see sub-suites (which can be composed into the parent suite or run as a standalone suite process) as an important tool to solve many suite design problems.

For makefile-esque, fire and forget suites we would ideally be able to run the cylc "kernel" in a lightweight configuration which has less impact on the filesystem.

I've very briefly recorded some of the core ideas here: https://github.com/cylc/cylc/wiki/Possible-Cylc-Futures#hierarchical-suites

**Comment by matthewrmshin** [comment]
https://github.com/metomi/rose/blob/master/lib/python/rose/job_runner.py is a Makefile-like system for running pre-defined tasks with dependencies in a multiprocessing pool. It is very usable on its own, and if we take out the dependency on `rose.reporter`, it should only import from modules in the standard library. I can polish it a bit, (perhaps relocate its logic), and add some examples so it can be used outside of Rose.

**Comment by hjoliver** [comment]
@TomekTrzeciak -

> Imagine a multitude of suites running concurrently and interacting with each other (via messages or data flow of some kind), but without a central coordinating authority,...

We can already have a multiple suites with cross-suite triggering via messages, although tightly coupled if using the built-in other-suite DB polling mechanism for cross-suite triggering. In the new production system in development at BoM, suites publish "data availability" messages to Kafka (message broker), and dependent suites, if interested, use the new external trigger mechanism to look for these messages in the right Kafka topic.  And Kafka can extend across zones (if allowed to, obviously); not sure about cross-site.  Of course each individual suite in this system is a normal suite as per the current incarnation of Cylc (with full run directory etc.) - do you envisage something more than this?  (many of these distributed suites running in the aforementioned lighter-weight make-like mode, presumably?).

**Comment by hjoliver** [comment]
@matthewrmshin - I wasn't aware of job_runner in Rose (and have not actually looked at yet, right now). It sounds like just what's needed for more sophisticated job bunching (i.e. with non-trivial dependencies). @TomekTrzeciak - are you guys using this?

**Comment by matthewrmshin** [comment]
Believe it or not, `rose.job_runner` was written to handle file installation in Rose (so everyone has been using it implicitly!) Unfortunately, I don't think it is used elsewhere.

(For the record, the logic is actually a Python port  + simplified version of what is inside of `fcm make` build system.)

**Comment by hjoliver** [comment]
Interesting!

**Comment by hjoliver** [comment]
>  not allowed in operations at our site

Huh, I'd be interested to know why that is (by email if you like).  We use it extensively, with no problems.

> I much prefer the publisher/subscriber pattern like the one you mentioned with xtriggers and Kafka, but currently this relies on an external system (Kafka) that is not part of Cylc.

Agreed, publisher/subscriber is much nicer.  Re-inventing that wheel (message broker) as a built-in capability is unlikely to be an option though, unless it turned out to be quite easy in the restricted context of Cylc usage.  (Interestingly we will have a central communications gateway - the "reverse proxy server" in the new web architecture...).

I think we will be going at least some way toward the architecture you describe, once we implement the "mean-and-lean execution engine with low start-up and overhead".

**Comment by TomekTrzeciak** [comment]
@hjoliver, come to think about it, xtriggers are still a polling interface. I think that cylc message + external triggers are more in spirit of the publisher/subscriber model (the notification part of the problem). What is still missing is the subscription part, i.e., there is no mechanism to request a notification about a specific output (event). With that added one could have a true publisher/subscriber interface.

(@cylc/core, do you think there should be a separate issue to discuss inter-suite communication and distributed running? I feel that this is quite important and fundamental aspect of running workflows that deserves some further discussion.)

About lack of inter-suite triggering at our site, I'm not sure about the exact reasons for this policy (@matthewrmshin or @oliver-sanders might know more), but one problem with it is that you need access to the suite database. This is not a problem if you run suites within the same security domain, but if you need to feed your dev suite from the operational one (e.g., testing prior to release), this doesn't work.

**Comment by hjoliver** [comment]
@TomekTrzeciak - yes xtriggers are still a polling interface, but if all xtriggers in all suites are looking at a central message broker (rather than targetting specific suites) at least there's no tight coupling. 

Note also #2477 (which needs updating - I don't think "persistent consumers" (as an alternative to polling) would be hard to implement.

> do you think there should be a separate issue to discuss inter-suite communication and distributed running?

Fine with me. However, I am concerned about the amount of effort to reinvent the message broker wheel unnecessarily (particular if persistent message consumers would obviate polling concerns). There's a number of message brokers out there, and some are very easy to use.  But maybe you can convince us this is different?

> one problem with it is that you need access to the suite database ...

That's exactly why we're using Kafka at BoM (well, there are other reasons to, such as sending info to a central reporting DB): not only do you not need to know where suite DB are, you don't even need to be in the same security domain, so long as you can see the message broker.

**Comment by hjoliver** [comment]
(Note xtriggers are not primarily intended for inter-suite triggering. Polling is the appropriate action for some kinds of interaction with the external world: when you're waiting for something to happen, but that something does not have any alert system that you can "subscribe" to).

**Comment by hjoliver** [comment]
This issue (as per title and description above) will be closed by PR #2759.

I've opened #2797 for official support of local run-dirs for sub-suites as discussed above.

And #2798 for  built-in support for inter-suite triggering via a message broker or similar, aso discussed above.
